### PR TITLE
Improved screenshot sync

### PR DESCRIPTION
**Pull request by lacostej**
For reviewing purposes.

See also comments on https://github.com/daniszewskik/deliver/commit/ef399609e79481837829cbf5dad5fb207777511a and https://github.com/lacostej/deliver/pull/1

### CODE REVIEW

File: deliver/lib/deliver/upload_screenshots.rb [hunk 1]
```diff
@@ -9,17 +9,44 @@ def upload(options, screenshots)
...
-      # First, clear all previously uploaded screenshots, but only where we have new ones
-      # screenshots.each do |screenshot|
-      #   to_remove = v.screenshots[screenshot.language].find_all do |current|
-      #     current.device_type == screenshot.device_type
-      #   end
-      #   to_remove.each { |t| t.reset! }
-      # end
+      v.screenshots.each do |lang, screenshots_for_lang|
+        screenshots_for_lang.each do |current|
+          matched = current.original_file_name.match(/ftl_[0-9a-f]{32}_*/)
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: you can use `/ftl_([0-9a-f]{32})_(*)/` and then `matched[1]`

File: deliver/lib/deliver/upload_screenshots.rb [hunk 2]
```diff
@@ -9,17 +9,44 @@ def upload(options, screenshots)
...
+          # store remote checksum. We will need it later to determine if we have to upload screenshot
+          @checksums_remote[current.language] ||= {}
+          @checksums_remote[current.language][current.device_type] ||= []
+          @checksums_remote[current.language][current.device_type] << checksum
+
+          # Remove from ITC non existing locally screenshots
+          # Or screenshots that have wrong order (compare indexes)
+          # Or screenshots that we haven't uploaded through spaceship
+          unless checksum && local_and_remote_match_to_delete(current, checksum)
+            UI.message("Deleting screenshot #{current.original_file_name} for language #{current.language}")
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: not sure if original_file_name helps (especially if it has the ftl_..._ helps).

Maybe do something like 

```
original_file_name = current.original_file_name
matched = original_file_name.match(/ftl_([0-9a-f]{32})_(*)/)
if matched 
  checksum = matched[1]
  original_file_name = matched[2]
end
...

  UI.message("Deleting screenshot ##{current.sort_order} #{original_file_name} ...."
```

File: deliver/lib/deliver/upload_screenshots.rb [hunk 3]
```diff
@@ -94,5 +130,19 @@ def collect_screenshots_for_languages(path)
 
       return screenshots
     end
+
+    private
+
+    def local_and_remote_match_to_delete(current, checksum)
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: use checksum or md5 consistently

File: deliver/lib/deliver/upload_screenshots.rb [hunk 4]
```diff
@@ -94,5 +130,19 @@ def collect_screenshots_for_languages(path)
...
+      @checksums_local[current.language][current.device_type] &&
+      @checksums_local[current.language][current.device_type].include?(checksum) &&
+      @checksums_local[current.language][current.device_type].index(checksum) + 1 == current.sort_order
+    end
+
+    def local_and_remote_match_to_skip(screenshot, device_type, index, md5)
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: the 2 functions are very very similar !

File: spaceship/lib/spaceship/du/utilities.rb [hunk 5]
```diff
@@ -70,6 +70,11 @@ def video_resolution(video_path)
       [res[1].to_i, res[2].to_i]
     end
 
-    module_function :content_type, :grab_video_preview, :portrait?, :resolution, :video_resolution
+    # returns md5 checksum of given screenshot
```
Comment (lacostej) [inline spaceship/lib/spaceship/du/utilities.rb]: use `@return` instead of returns

File: spaceship/lib/spaceship/du/utilities.rb [hunk 6]
```diff
@@ -70,6 +70,11 @@ def video_resolution(video_path)
...
+    def md5digest(screenshot_path)
```
Comment (lacostej) [inline spaceship/lib/spaceship/du/utilities.rb]: rename screenshot_path with `file_path`

File: deliver/lib/deliver/upload_screenshots.rb [hunk 7]
```diff
@@ -94,5 +130,19 @@ def collect_screenshots_for_languages(path)
...
+      @checksums_remote[screenshot.language][device_type] &&
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: I would have introduced the following class:

```
# used to maintain 2 set of MD5s one for local one for remotes
class ScreenshotMD5s
  def initialize
    @checksums = {}
  end

  # index is implicitly computed
  def add_md5(language, device_type, md5)
    @checksums[language] ||= {}
    @checksums[language][device_type] ||= []
    @checksums[language][device_type] << md5
  end

  def matches_md5(language, device_type, md5, index)
    @checksums[language][device_type] &&
      @checksums[language][device_type].include?(md5) &&
      @checksums[language][device_type].index(md5) + 1 == index
   end
end
```

and refactor the whole thing.

File: deliver/spec/upload_screenshots_spec.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      File.open(file_path, 'w') { |file| file.write("#{i}") }
+    end
+  end
+
+  def mock_local_upload(sequence)
+    @order = 0
+    to_upload = sequence.map do |i|
+      file_path = File.join(@root, "scr_#{i}.jpg")
+      md5 = Spaceship::Utilities.md5digest(file_path)
+      file = Screenshot.new({file_path: file_path, language: 'en-US', original_file_name: "ftl_#{md5}_scr_#{i}.jpg", sort_order: @order})
```
Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb:23]: doesn't order start at 1 ?

File: deliver/spec/upload_screenshots_spec.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    to_upload
+  end
+
+  let(:options) { { app: FakeApp.new } }
+
+  context "Deleting screenshots from ITC" do
+    it "should delete screenshot from ITC when order had been changed" do
+      # 1st and 2nd changed places. Expecting deletion of both
+      expect do
+        Deliver::UploadScreenshots.new.upload(options, mock_local_upload([2, 1, 3, 4]))
```
Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb]: do we need to create a new UploadScreenshots instance every time ? Let's test with the same, to make sure it doesn't keep state that it shouldn't. (I am not so sure of the way the @local_checksums are kept. Maybe they should be passed around instead. Less state is sometimes better.

File: deliver/spec/upload_screenshots_spec.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      end.to output(/Deleting 1 for device iphone4\nDeleting 2 for device iphone4/).to_stdout
```
Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb]: what is the order at the end ?, 2, 1, 3, 4? are we checking that somewhere ?

File: deliver/spec/upload_screenshots_spec.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    it "should delete screenshot from ITC when file had been changed" do
+       # 1st changed. Expecting deletion of first
+      expect do
+        Deliver::UploadScreenshots.new.upload(options, mock_local_upload([5, 2, 3, 4]))
+      end.to output(/Deleting 1 for device iphone4/).to_stdout
+    end
+
+    it "should delete screenshot from ITC when there's no md5 in filename" do
+
+      # add screenshot without md5
```
Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb]: (aka "simulate screenshot not uploaded by spaceship")

File: deliver/spec/upload_screenshots_spec.rb [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      file_path = File.join(@root, "scr_5.jpg")
+      file = Screenshot.new({file_path: file_path, language: 'en-US', original_file_name: "scr_5.jpg", sort_order: 5})
+      options[:app].edit_version.screenshots['en-US'] << file
+
+
+      # 5th don't have md5 in file name. Expecting deletion
```
Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb]: deletion and re-upload. Do we test for both ?

File: deliver/spec/upload_screenshots_spec.rb (hunk 8 above)
Comment (daniszewskik) [inline deliver/spec/upload_screenshots_spec.rb:23]: Apple ITCScreenshot starts form 1. Local upload starts normally form 0, so it's closer to real life

Comment (lacostej) [inline deliver/spec/upload_screenshots_spec.rb:23]: ah yes. I thought they had mirrored both.

File: deliver/lib/deliver/upload_screenshots.rb [hunk 13]
```diff
@@ -1,4 +1,25 @@
...
+    end
+
+    # index is implicitly computed
+    def add_md5(language, device_type, md5)
+      @checksums[language] ||= {}
+      @checksums[language][device_type] ||= []
+      @checksums[language][device_type] << md5
+    end
+
+    def matches_md5(language, device_type, md5, index)
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: small nitpick rename to `matches_md5?`

File: deliver/lib/deliver/upload_screenshots.rb [hunk 14]
```diff
@@ -9,17 +30,41 @@ def upload(options, screenshots)
       v = app.edit_version
       UI.user_error!("Could not find a version to edit for app '#{app.name}'") unless v
 
-      UI.message("Starting with the upload of screenshots...")
+      # md5's of uploaded screenshots
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: this comment is slightly incorrect. It' the md5 of "screenshots to upload"

we could get rid of these comments anyway

File: deliver/lib/deliver/upload_screenshots.rb [hunk 15]
```diff
@@ -9,17 +30,41 @@ def upload(options, screenshots)
...
+      @checksums_local = ScreenshotMD5s.new
+      # md5's of ITC screenshots
+      @checksums_remote = ScreenshotMD5s.new
+
+      screenshots.each do |screenshot|
+        # checksum of uploaded screenshot
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: same here

**Comment by lacostej** [comment]
just address the things related to the comments, and squash ! Can't wait to test :)

File: deliver/lib/deliver/upload_screenshots.rb [hunk 16]
```diff
@@ -9,23 +31,52 @@ def upload(options, screenshots)
...
+            md5 = matched[1]
+            original_file_name = matched[2]
+          end
+          # store remote checksum. We will need it later to determine if we have to upload screenshot
+          @checksums_remote.add_md5(current.language, current.device_type, md5)
+          # Remove from ITC non existing locally screenshots
+          # Or screenshots that have wrong order (compare indexes)
+          # Or screenshots that we haven't uploaded through spaceship
+          unless md5 && @checksums_local.matches_md5?(current.language, current.device_type, md5, current.sort_order)
+            UI.message("Deleting screenshot #{original_file_name} for language #{current.language}")
```
Comment (lacostej) [inline deliver/lib/deliver/upload_screenshots.rb]: we missed that one

I would add here `##{current.sort_order}` (to display the index)

**Pull request closed without merging**